packages/jest-mock/src/index.ts

```diff
@@ -1008,40 +1008,48 @@ export class ModuleMocker {
     return fn;
   }
 
-  spyOn<T extends object, M extends PropertyLikeKeys<T>>(
+  spyOn<
+    T extends object,
+    K extends PropertyLikeKeys<Required<T>>,
+    V extends Required<T>[K],
+  >(object: T, methodKey: K, accessType: 'get'): SpyInstance<() => V>;
+
+  spyOn<
+    T extends object,
+    K extends PropertyLikeKeys<Required<T>>,
+    V extends Required<T>[K],
+  >(object: T, methodKey: K, accessType: 'set'): SpyInstance<(arg: V) => void>;
+
+  spyOn<
+    T extends object,
+    K extends ConstructorLikeKeys<Required<T>>,
+    V extends Required<T>[K],
```

Comment:
`V` is just a helper to avoid `SpyInstance<(...args: ConstructorParameters<Required<T>[K]>) => InstanceType<Required<T>[K]>>`